infrastructure/src/main/java/org/corfudb/infrastructure/CompactorLeaderServices.java

```diff
@@ -0,0 +1,494 @@
+            return LeaderServicesStatus.FAIL;
+        }
+        return LeaderServicesStatus.SUCCESS;
+    }
+
+    /**
+     * Validates the liveness of any on-going checkpoint of a table
+     * ActiveCheckpointTable contains the list of tables for which checkpointing has started. This method is scheduled
+     * to execute continuously by the leader, which monitors the checkpoint activity of the tables present in
+     * ActiveCheckpointTable.
+     * if there are no tables present,
+     * ... check the CheckpointStatusTable for any progress made (this step is done to include the
+     * tables which are checkpointed really fast)
+     * ... if there's no progress for timeout ms, call finishCompactionCycle() to mark the end of the cycle
+     * if there are any slow checkpointers,
+     * ... monitor checkpointing of the table by observing if the checkpointStream's tail moves forward
+     * ... if it does not move forward for timeout ms, then mark it as failed
+     * Also, when checkpoint of a table is found to be failed, the cycle is immediately marks as failed.
+     *
+     * @param timeout in ms
+     */
+    public void validateLiveness(long timeout) {
+        List<TableName> tableNames;
+        try (TxnContext txn = corfuStore.txn(CORFU_SYSTEM_NAMESPACE)) {
+            tableNames = txn.keySet(activeCheckpointsTable).stream().collect(Collectors.toList());
+            txn.commit();
+        } catch (Exception e) {
+            syslog.warn("Unable to acquire tableNames");
+            return;
+        }
+
+        long currentTime = System.currentTimeMillis();
+        if (tableNames.isEmpty()) {
+            handleNoActiveCheckpointers(timeout, currentTime);
+        }
+        for (TableName table : tableNames) {
+            livenessValidatorHelper.clear();
+            String streamName = TableRegistry.getFullyQualifiedTableName(table.getNamespace(), table.getTableName());
+            UUID streamId = CorfuRuntime.getCheckpointStreamIdFromName(streamName);
+            long currentStreamTail = corfuRuntime.getSequencerView()
+                    .getStreamAddressSpace(new StreamAddressRange(streamId, Address.MAX, Address.NON_ADDRESS)).getTail();
+
+            long syncHeartBeat = LIVENESS_INIT_VALUE;
+            try (TxnContext txn = corfuStore.txn(CORFU_SYSTEM_NAMESPACE)) {
+                ActiveCPStreamMsg activeCPStreamMsg = txn.getRecord(activeCheckpointsTable, table).getPayload();
+                if (activeCPStreamMsg == null) {
+                    continue;
+                }
+                syncHeartBeat = activeCPStreamMsg.getSyncHeartbeat();
+                txn.commit();
+            } catch (Exception e) {
+                syslog.warn("Unable to acquire table status", e);
+            }
+            if (readCache.containsKey(table)) {
+                LivenessMetadata previousStatus = readCache.get(table);
+                if (previousStatus.getStreamTail() < currentStreamTail) {
```

Comment:
In case of a sequencer regression  the delta |currentStreamTail - previousStatus.getStreamTail| can be so huge that the readCache is never updated.